### ISSUE TITLE
Exports for Comfyui?

### ISSUE THREAD

**Issue by Jules-A** [Artoriuz/ArtCNN#47]
I haven't managed to get Vaporsynth running properly on my system so I wasn't able to properly test your newer variants. (Every single time there would be issues, corruption or just slower than glsl) and have given up on Vaporsynth for good.

However, I decided to try them in Comfyui to see how they'd fare upscaling there. I couldn't find any existing nodes or a better way to run them in there so I just quickly vibe coded a node to run it https://github.com/Jules-A/ComfyUI-ONNX_ArtCNN and honestly rather impressed even the 4x RGB GAN can upscale from 1MP in ~500ms on my 9070XT even in DirectML backend...

It's been hell trying to find an upscaler for Comfyui that's fast as most of the smaller models have significant issues, is it possible to get a .pth/safetensor export, maybe at least for the bigger models?

Also, unrelated but I heard ffmpeg has a DNN filter now, I haven't really looked into it, have you?

**Comment by Artoriuz** [comment]
To upscale images, you can just call ORT directly or use the [included inferencer](https://github.com/Artoriuz/ArtCNN/tree/main/Inferencer).

The reason why I don't make GLSL exports of the bigger models is that the resulting shaders are too slow. Yes, as you've observed the actual inference engines are often much faster. As far as I'm aware, there's no straightforward way to use ORT on mpv, the workaround is to use it through vsmlrt in a vapoursynth script.

As for ComfyUI integration, I haven't looked into it as I don't use ComfyUI. However, if there's a way to call ORT it should just work. Just make sure the arrays you're giving the model are in the expected [0.0, 1.0] range and in the NCHW format.

**Comment by Jules-A** [comment]
> As for ComfyUI integration, I haven't looked into it as I don't use ComfyUI. However, if there's a way to call ORT it should just work. Just make sure the arrays you're giving the model are in the expected [0.0, 1.0] range and in the NCHW format.

That's what I'm already doing but not through the included script as that would encode it outside and make displaying a preview annoying and possibly slower.

`  
  
        ycbcr = _rgb_to_ycbcr(rgb_nchw)
        y = ycbcr[:, 0:1]
        y_np = y.cpu().numpy().astype(np.float32, copy=False)
        y2_np = _run_luma(luma_session, y_np)
        y2 = torch.from_numpy(y2_np)

        in_h, in_w = y.shape[-2:]
        out_h, out_w = y2.shape[-2:]
        scale_h = out_h / max(in_h, 1)
        scale_w = out_w / max(in_w, 1)
        if abs(scale_h - scale_w) > 1e-6:
            raise ArtCNNONNXError(
                "Luma ArtCNN model produced a non-uniform scale: "
                f"Input={in_h}x{in_w}, output={out_h}x{out_w} "
                f"({scale_w:.3f}x{scale_h:.3f})"
            )
        if min(abs(scale_w - 2.0), abs(scale_w - 4.0)) > 1e-6:
            raise ArtCNNONNXError(
                "This node supports 2x and 4x luma ArtCNN models. "
                f"Input={in_h}x{in_w}, output={out_h}x{out_w} "
                f"({scale_w:.3f}x{scale_h:.3f})"
            )

        # ArtCNN chroma models explicitly require bilinear chroma pre-upscaling.
        # When no neural chroma model is selected, allow a conventional resampler.
        if chroma_session is not None:
            cbcr_up = _resize_chroma(ycbcr[:, 1:3], (out_h, out_w), "bilinear")
            chroma_in = torch.cat((y2, cbcr_up), dim=1)
            chroma_np = chroma_in.cpu().numpy().astype(np.float32, copy=False)
            chroma2_np = _run_chroma(chroma_session, chroma_np)
            chroma2 = torch.from_numpy(chroma2_np).clamp(0, 1)
        else:
            # For ordinary RGB input there is no subsampled chroma plane to
            # independently reconstruct.  Build the output chroma from a
            # spatially upscaled RGB reference at the exact output geometry,
            # then replace only its Y plane with ArtCNN's result. This keeps
            # chroma geometrically aligned with the ArtCNN luma and avoids
            # color casts from independently resampling Cb/Cr.
            if chroma_resampler.lower() == "bilinear":
                rgb_up = F.interpolate(
                    rgb_nchw, size=(out_h, out_w), mode="bilinear", align_corners=False
                )
            else:
                # Reuse the deterministic reconstruction kernels for each RGB
                # channel. This is slower than bilinear but preserves the
                # requested filter without converting color planes separately.
                rgb_up_np = rgb_nchw.detach().cpu().numpy().astype(np.float32, copy=False)
                rgb_up_np = _resize_axis_np(rgb_up_np, out_h, axis=2, kind=chroma_resampler.lower())
                rgb_up_np = _resize_axis_np(rgb_up_np, out_w, axis=3, kind=chroma_resampler.lower())
                rgb_up = torch.from_numpy(rgb_up_np)
            chroma2 = _rgb_to_ycbcr(rgb_up)[:, 1:3].clamp(0, 1)

        out_ycbcr = torch.cat((y2.clamp(0, 1), chroma2), dim=1)
        out = _ycbcr_to_rgb(out_ycbcr).permute(0, 2, 3, 1).contiguous()

        if alpha is not None:
            alpha_up = F.interpolate(
                alpha.permute(0, 3, 1, 2),
                size=(out_h, out_w),
                mode="nearest",
            ).permute(0, 2, 3, 1)
            out = torch.cat((out, alpha_up), dim=-1)

        return (out,)

`


it seems to be clamping and in NCHW format. I was just hoping to have it a comfy compatible format so it can take a possibly more optmized path as I don't think ORT is the fastest backend on AMD. Plus also I'm not sure if the AI is doing it correctly as I'm unfamilar with torch and pretty much python in general.

**Comment by Artoriuz** [comment]
MIGraphX is AMD's first-party inference engine an and it can be used as an ONNX Runtime Execution Provider.

Torch is mostly meant for training and it's also known for being slow. The actual inference engines should all be faster, including DML.

**Comment by Jules-A** [comment]
> MIGraphX is AMD's first-party inference engine an and it can be used as an ONNX Runtime Execution Provider.
> 
> Torch is mostly meant for training and it's also known for being slow. The actual inference engines should all be faster, including DML.

I've never managed to get the MIGraphx version to work in MPV unfortunately. I asked the AI to try use it with the package but it's stuck trying to compile the model to mxr for ~40mins so far and I forgot why that was even needed but it failed to interface directly through vsmlrt
